File: bcryptPasswordInExcel.js
const fs = require('fs');
const bcrypt = require('bcrypt');
const XLSX = require('xlsx');

// Fungsi untuk mengenkripsi password
function hashPassword(password) {
  const salt = bcrypt.genSaltSync(12);
  const hashedPassword = bcrypt.hashSync(password, salt);
  return hashedPassword;
}

// Fungsi untuk memproses setiap baris pada file Excel
function processRow(row) {
  const updatedRow = { ...row };
  updatedRow.Password = hashPassword(row.Password);
  console.log("Password:", updatedRow.Password);
  return updatedRow;
}

// Fungsi utama
function main() {
  const workbook = XLSX.readFile('Dealer.xlsx');

  const worksheet = workbook.Sheets[workbook.SheetNames[0]];
  const jsonData = XLSX.utils.sheet_to_json(worksheet);

  const processedData = jsonData.map(processRow);

  const newWorkbook = XLSX.utils.book_new();
  const newWorksheet = XLSX.utils.json_to_sheet(processedData);
  XLSX.utils.book_append_sheet(newWorkbook, newWorksheet, 'Sheet1');
  XLSX.writeFile(newWorkbook, 'Dealer.xlsx');
}

main();

File: bcryptPasswordInExcel.go
package main

import (
	"fmt"
	"log"
	"os"
	"sync"

	"github.com/tealeg/xlsx"
	"golang.org/x/crypto/bcrypt"
)

// Worker pool
type Worker struct {
	ID        int
	JobQueue  chan *xlsx.Row
	ResultMap map[int]*xlsx.Row
	Wg        *sync.WaitGroup
}

// Fungsi untuk mengenkripsi password
func hashPassword(password string) string {
	hashedPassword, err := bcrypt.GenerateFromPassword([]byte(password), bcrypt.DefaultCost)
	if err != nil {
		log.Fatal(err)
	}
	return string(hashedPassword)
}

// Fungsi untuk memproses setiap baris pada file Excel
func processRow(row *xlsx.Row) *xlsx.Row {
	cells := row.Cells

	passwordCell := cells[1]
	password := passwordCell.String()

	hashedPassword := hashPassword(password)
	passwordCell.SetString(hashedPassword)

	return row
}

// Fungsi untuk menjalankan pekerjaan pemrosesan pada worker
func (w *Worker) Process() {
	for row := range w.JobQueue {
		processedRow := processRow(row)
		w.ResultMap[row.GetCoordinate()] = processedRow
		w.Wg.Done()
	}
}

// Fungsi utama
func main() {
	excelFileName := "Dealer.xlsx"

	// Baca file Excel
	xlFile, err := xlsx.OpenFile(excelFileName)
	if err != nil {
		log.Fatal(err)
	}

	// Dapatkan sheet pertama dari file Excel
	sheet := xlFile.Sheets[0]

	// Buat job queue dan result map
	jobQueue := make(chan *xlsx.Row, len(sheet.Rows))
	resultMap := make(map[int]*xlsx.Row)

	// Buat worker pool
	workerCount := 4 // Jumlah worker yang diinginkan
	wg := &sync.WaitGroup{}
	wg.Add(len(sheet.Rows))
	workers := make([]*Worker, workerCount)

	for i := 0; i < workerCount; i++ {
		worker := &Worker{
			ID:        i,
			JobQueue:  jobQueue,
			ResultMap: resultMap,
			Wg:        wg,
		}
		workers[i] = worker
		go worker.Process()
	}

	// Tambahkan pekerjaan ke job queue
	for _, row := range sheet.Rows {
		jobQueue <- row
	}

	// Tutup job queue
	close(jobQueue)

	// Tunggu pemrosesan selesai
	wg.Wait()

	// Tulis kembali ke file Excel
	outputFileName := "Dealer.xlsx"
	for _, row := range resultMap {
		err = row.Sheet().Row(row.RowIndex).ReplaceRow(row)
		if err != nil {
			log.Fatal(err)
		}
	}

	err = xlFile.Save(outputFileName)
	if err != nil {
		log.Fatal(err)
	}

	fmt.Println("File Excel berhasil diubah.")
}
